cpp/src/spatial_window/spatial_window.cu

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_view.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/copy_if.cuh>
+
+#include <cuspatial/error.hpp>
+
+#include <memory>
+#include <type_traits>
+
+namespace {
+
+// Functor to filter out points that are not inside the query window
+// This is passed to cudf::detail::copy_if
+template <typename T>
+struct spatial_window_filter {
+  spatial_window_filter(T window_min_x,
+                        T window_max_x,
+                        T window_min_y,
+                        T window_max_y,
+                        cudf::column_device_view const& x,
+                        cudf::column_device_view const& y)
+    : min_x{std::min(window_min_x, window_max_x)},  // support mirrored rectangles
```

Comment:
I'm on the fence about supporting mirrored windows.

There is a case where keeping the values as-is becomes a feature in and of itself. If an algorithm upstream _intentionally_ produces a "negative-size" window, that is an implicit zero-results query we can optimize for.

On the one hand, a user might misunderstand the API and wonder why no results are being returned.
On the other hand, a user might provide a "negative-size" or "too small" window size intentionally, only to have us rectify it, and wonder why _any_ results are returning at all.

If we rectify the window, we actually lose flexibility in how the API can be used (and complicate the documentation, since there are now 4 "quadrants" for rectifiable window coordinates).

Reply:
Alternatively, we could move the rectify to the python layer and make optional.

Reply:
The previous behavior was to return an error, rather than zero results.

Reply:
I'm on the fence too, so we'll just leave it until someone files an issue. :)

Reply:
Updated docs to reflect real behavior.